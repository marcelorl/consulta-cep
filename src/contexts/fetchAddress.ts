import React from 'react'

import { fetchAddressByZipCode, fetchMapCoordinates } from '../services/address'

export default async (dispatch: any, zipCode: string) => {
  try {
    const addressData = await fetchAddressByZipCode(zipCode)
    const mapsCoordinates = await fetchMapCoordinates(zipCode)

    return dispatch({
      type: 'FETCH_ADDRESS',
      payload: {...addressData, mapsCoordinates}
    })
  } catch (e) {
    return dispatch({
      type: 'FETCH_ADDRESS_FAIL'
    })
  }
}
