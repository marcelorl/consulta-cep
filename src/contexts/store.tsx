import React from 'react'

import {
  IState,
  IContextProps,
  Actions
} from '../models'
import { fetchAddressByZipCode, fetchMapCoordinates } from '../services/address'

export const Store = React.createContext({} as IContextProps)

const initialState: IState = {
  address: {},
  status: ''
}

const reducer = (state: IState, action: Actions) => {
  switch (action.type) {
    case 'FETCH_ADDRESS':
      return { ...state, address: action.payload, status: 'LOADED' }
    case 'FETCH_ADDRESS_FAIL':
      return { ...state, address: {}, status: 'ERROR' }
    case 'ADDRESS_RESET': {
      return { ...state, address: {}, status: '' }
    }
    default:
      return state
  }
}

export const useAddress = () => {
  const [ state, dispatch ] = React.useReducer(reducer, initialState)

  const fetchAddress = async (zipCode: string) => {
    try {
      const addressData = await fetchAddressByZipCode(zipCode)
      const mapsCoordinates = await fetchMapCoordinates(addressData)

      return dispatch({
        type: 'FETCH_ADDRESS',
        payload: {...addressData, mapsCoordinates}
      })
    } catch (e) {
      return dispatch({
        type: 'FETCH_ADDRESS_FAIL'
      })
    }
  }

  const resetState = () =>
    dispatch({
      type: 'ADDRESS_RESET'
    })

  return [ state, { fetchAddress, resetState } ]
}

export const StoreProvider = (props: any) => {
  const [ state, action ] = useAddress()

  return (
    <Store.Provider value={{ state, action }}>
      {props.children}
    </Store.Provider>
  )
}
