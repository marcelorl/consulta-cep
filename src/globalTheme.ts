import { DefaultTheme } from 'styled-components'

export const globalTheme: DefaultTheme = {
  colors: {
    darkgray: '#999999',
    gainsboro: '#e0ded9',
    lightblue: '#aadaff',
    white: '#ffffff',
  }
}
