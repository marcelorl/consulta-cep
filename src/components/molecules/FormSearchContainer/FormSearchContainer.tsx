import React from 'react'
import styled from 'styled-components'

import { Store } from '../../../contexts/store'
import Button from '../../atoms/Button'
import SearchField from '../../atoms/SearchField'

const FormSearchContainerWrapper = styled.form`
  background: ${props => props.theme.colors.gainsboro};
  box-sizing: border-box;
  padding: 2rem 2rem 4rem;
  width: 100%;
`

const FormSearchFieldWrapperLabel = styled.div`
  font-weight: 600;
  margin-bottom: 2rem;
`

const SearchBar = styled.div`
  color: ${props => props.theme.colors.darkgray};
  display: flex;
`

const LabelWrapper = styled.label`
  > * {
    margin-right: 1rem;
  }
`

const FormSearchContainer = () => {
  const ctx = React.useContext(Store)
  const [ zipCode, setZipCode ] = React.useState('')

  const onSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()

    await ctx.action.fetchAddress(zipCode)
  }

  return (
    <FormSearchContainerWrapper onSubmit={onSubmit}>
      <FormSearchFieldWrapperLabel>Consultar</FormSearchFieldWrapperLabel>
      <SearchBar>
        <LabelWrapper htmlFor="search-field">
          <span>CEP</span>
          <SearchField id="search-field" value={zipCode} onChange={setZipCode} />
        </LabelWrapper>
        <Button />
      </SearchBar>
    </FormSearchContainerWrapper>
  )
}

export default React.memo(FormSearchContainer)
