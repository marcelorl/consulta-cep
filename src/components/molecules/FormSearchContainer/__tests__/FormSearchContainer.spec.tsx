import React from 'react'
import { ThemeProvider } from 'styled-components'
import { cleanup, fireEvent, render } from 'react-testing-library'

import FormSearchContainer from '../FormSearchContainer'
import { StoreProvider } from '../../../../contexts/store'
import { globalTheme } from '../../../../globalTheme'

afterEach(cleanup)

describe('#FormSearchContainer', () => {
  it('should render FormSearchContainer', () => {
    const { container } = render(
      <ThemeProvider theme={globalTheme}>
        <FormSearchContainer />
      </ThemeProvider>
    )

    expect(container).toMatchSnapshot()
  })

  it('should not call fetch, because zipcode is required', () => {
    const { getByText } = render(
      <StoreProvider>
        <ThemeProvider theme={globalTheme}>
          <FormSearchContainer />
        </ThemeProvider>
      </StoreProvider>
    )

    fireEvent.click(getByText(/buscar/i))

    expect(fetch).not.toBeCalled()
  })

  it('should call fetch, because zipcode is now filled', () => {
    const { getByText, getByTestId } = render(
      <StoreProvider>
        <ThemeProvider theme={globalTheme}>
          <FormSearchContainer />
        </ThemeProvider>
      </StoreProvider>
    )

    fireEvent.change(getByTestId('search-field'), { target: { value: '00000-000' } })
    fireEvent.click(getByText(/buscar/i))

    expect(fetch).toBeCalled()
  })
})


