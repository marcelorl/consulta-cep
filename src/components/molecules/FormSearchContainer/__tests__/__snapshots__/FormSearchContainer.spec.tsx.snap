// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#FormSearchContainer should render FormSearchContainer 1`] = `
.c5 {
  border-radius: 5px;
  background: #aadaff;
  color: #ffffff;
  cursor: pointer;
  font-size: 1rem;
  padding: .35rem 1rem;
}

.c5:hover {
  border: 1px solid #aadaff;
}

.c4 {
  border: 1px solid #999999;
  border-radius: 5px;
  box-shadow: 0;
  font-size: 1rem;
  padding: 5px 10px;
}

.c0 {
  background: #e0ded9;
  box-sizing: border-box;
  padding: 2rem 2rem 4rem;
  width: 100%;
}

.c1 {
  font-weight: 600;
  margin-bottom: 2rem;
}

.c2 {
  color: #999999;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 > * {
  margin-right: 1rem;
}

<div>
  <form
    class="c0"
    data-testid="form-search-container"
  >
    <div
      class="c1"
    >
      Consultar
    </div>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="search-field"
      >
        <span>
          CEP
        </span>
        <input
          class="c4"
          data-testid="search-field"
          id="search-field"
          placeholder="01234-567"
          required=""
          value=""
        />
      </label>
      <button
        class="c5"
      >
        Buscar
      </button>
    </div>
  </form>
</div>
`;
